interfaces/builtin/x11.go

```diff
@@ -60,14 +137,53 @@ socket AF_NETLINK - NETLINK_KOBJECT_UEVENT
 bind
 `
 
+type x11Interface struct {
+	commonInterface
+}
+
+func (iface *x11Interface) AppArmorConnectedSlot(spec *apparmor.Specification, plug *interfaces.ConnectedPlug, slot *interfaces.ConnectedSlot) error {
+	if !release.OnClassic {
+		old := "###PLUG_SECURITY_TAGS###"
+		new := plugAppLabelExpr(plug)
+		snippet := strings.Replace(x11ConnectedSlotAppArmor, old, new, -1)
+		spec.AddSnippet(snippet)
+	}
+	return nil
+}
+
+func (iface *x11Interface) SecCompPermanentSlot(spec *seccomp.Specification, slot *snap.SlotInfo) error {
+	if !release.OnClassic {
+		spec.AddSnippet(x11PermanentSlotSecComp)
+	}
+	return nil
+}
+
+func (iface *x11Interface) AppArmorPermanentSlot(spec *apparmor.Specification, slot *snap.SlotInfo) error {
+	if !release.OnClassic {
+		spec.AddSnippet(x11PermanentSlotAppArmor)
+	}
+	return nil
+}
+
+func (iface *x11Interface) UDevPermanentSlot(spec *udev.Specification, slot *snap.SlotInfo) error {
+	if !release.OnClassic {
+		spec.TagDevice(`KERNEL=="tty[0-9]*"`)
+		spec.TagDevice(`KERNEL=="mice"`)
+		spec.TagDevice(`KERNEL=="mouse[0-9]*"`)
+		spec.TagDevice(`KERNEL=="event[0-9]*"`)
+		spec.TagDevice(`KERNEL=="ts[0-9]*"`)
+	}
+	return nil
+}
+
 func init() {
-	registerIface(&commonInterface{
-		name:                  "x11",
-		summary:               x11Summary,
-		implicitOnClassic:     true,
-		baseDeclarationSlots:  x11BaseDeclarationSlots,
-		connectedPlugAppArmor: x11ConnectedPlugAppArmor,
-		connectedPlugSecComp:  x11ConnectedPlugSecComp,
-		reservedForOS:         true,
-	})
+	registerIface(&x11Interface{commonInterface{
+		name:                   "x11",
+		summary:                x11Summary,
+		implicitOnClassic:      true,
+		baseDeclarationSlots:   x11BaseDeclarationSlots,
+		connectedPlugAppArmor:  x11ConnectedPlugAppArmor,
+		connectedPlugSecComp:   x11ConnectedPlugSecComp,
```

Comment:
I know @zyga recommended commonInterface (which is fine), but note that this is the first slot implementation to use commonInterface. It's a little weird here for @gerboland to have to list the Plug stuff but not permanentSlotAppArmor, permanentSlotSecComp (btw, I hate that we camel-case SecComp, it is Seccomp), connectedSlotAppAmor, connectedSlotSecComp, permanentSlotUdev, etc.

@gerboland - this isn't for you to change unless @zyga changes his mind. I just found it hard to verify that enough was in this registerIface since commonInterface only implements some things.

Reply:
I'm still happy to see this pattern used as it makes code shorter! +1 on this.